opm/autodiff/ExtendedBlackoilModel_impl.hpp

```diff
@@ -0,0 +1,703 @@
+    void
+    ExtendedBlackoilModel<Grid>::computeAccum(const SolutionState& state,
+                                             const int            aix  )
+    {
+        Base::computeAccum(state, aix);
+
+        // Compute accumulation of the solvent
+        if (has_solvent_) {
+            const ADB& press = state.pressure;
+            const ADB& ss = state.solvent_saturation;
+            const ADB pv_mult = poroMult(press); // also computed in Base::computeAccum, could be optimized.
+            const Opm::PhaseUsage& pu = fluid_.phaseUsage();
+
+            const ADB& pg = state.canonical_phase_pressures[pu.phase_pos[Gas]];
+            rq_[solvent_pos_].b = solvent_props_.bSolvent(pg,cells_);
+            rq_[solvent_pos_].accum[aix] = pv_mult * rq_[solvent_pos_].b * ss;
+        }
+    }
+
+
+
+
+
+    template <class Grid>
+    void
+    ExtendedBlackoilModel<Grid>::
+    assembleMassBalanceEq(const SolutionState& state)
+    {
+
+        Base::assembleMassBalanceEq(state);
+
+        if (has_solvent_) {
+            residual_.material_balance_eq[ solvent_pos_ ] =
+                pvdt_ * (rq_[solvent_pos_].accum[1] - rq_[solvent_pos_].accum[0])
+                + ops_.div*rq_[solvent_pos_].mflux;
+        }
+
+    }
+
+
+
+
+
+    template <class Grid>
+    void ExtendedBlackoilModel<Grid>::addWellContributionToMassBalanceEq(std::vector<ADB>& cq_s,
+                                                                        const SolutionState& state,
+                                                                        WellState& xw)
+
+    {
+
+        // Add well contributions to solvent mass balance equation
+
+        Base::addWellContributionToMassBalanceEq(cq_s, state, xw);
+
+        if (has_solvent_) {
+            const int nperf = wells().well_connpos[wells().number_of_wells];
+            const int nc = Opm::AutoDiffGrid::numCells(grid_);
+
+            const Opm::PhaseUsage& pu = fluid_.phaseUsage();
+            const ADB zero = ADB::constant(V::Zero(nc));
+            const ADB& ss = state.solvent_saturation;
+            const ADB& sg = (active_[ Gas ]
+                             ? state.saturation[ pu.phase_pos[ Gas ] ]
+                             : zero);
+
+            Selector<double> zero_selector(ss.value(), Selector<double>::Zero);
+            V F_solvent = zero_selector.select(ss, ss / (ss + sg)).value();
+
+            const std::vector<int> well_cells(wells().well_cells, wells().well_cells + nperf);
+            const int nw = wells().number_of_wells;
+            V wellSolventFraction = Eigen::Map<const V>(&xw.solventFraction()[0], nperf);
+
+            for (int w = 0; w < nw; ++w) {
+                if(wells().type[w] == PRODUCER) {
+                    for (int perf = wells().well_connpos[w]; perf < wells().well_connpos[w+1]; ++perf) {
+                        wellSolventFraction[perf] = F_solvent[well_cells[perf]];
+                    }
+                }
+            }
+
```

Comment:
This part is a little hard to understand, what do the equations look like for the well flow?